### ISSUE TITLE
internal::ImplicitCast_() changes pointer value

### ISSUE THREAD

**Issue by schoolo** [google/googletest#889]
I am compiling using VisualStudio 2013 version 12.0.40629.00 Update 5.  I noticed that when I wrapped one of my mock objects in `NiceMock<>`, my tests started failing.  I dug in and realized that the ImplicitCast_ called in NiceMock's construct was actually _changing the value_ of the pointer that got passed into it.  I'm not sure this is your problem (seems like a compiler bug), but I figured I'd throw it up here anyway in case you guys wanted to use the information.  Screenshot attached.
<img width="957" alt="vs2013_compiler_bug_gmock_markedup" src="https://cloud.githubusercontent.com/assets/2209291/18896088/6f6525e6-84e4-11e6-938b-5b983ea119fd.PNG">

**Comment by schoolo** [comment]
BTW, I noticed this causing an issue because my tests depend on using 		`Mock::VerifyAndClearExpectations(void *)` , which takes an externally supplied pointer instead of the one associated with the mock object.  The externally supplied pointer (the real address) didn't match up with the address that had been stored (the one mutated by `testing::internal::ImplicitCast_()`)

**Comment by Bu11etmagnet** [comment]
@schoolo 
* Please don't post pictures of code. Use your editor's copy & paste commands.
* `ImplicitCast_` works like `static_cast`. When you `static_cast` from a derived class to a base class, the pointer value is not necessarily the same (think of multiple inheritance. If you static_cast  a `struct D : B1, B2` to a `B2`, the pointer has to be different). If you rely on the binary value of the pointer being the same, you are doing something wrong.

```
#include <iostream>
struct B1 { int b1_; };
struct B2 { int b2_; };
struct D : B1 , B2 { int d_; };
int main()
{
        D d;
        std::cout << "D =" << &d << "\nB1=" << static_cast<B1*>(&d) << "\nB2=" << static_cast<B2*>(&d) << "\n";
}
```
The results were similar with every compiler I could get my hands on:
```
$ ./a.out 
D =0x7ffd4a7a0940
B1=0x7ffd4a7a0940
B2=0x7ffd4a7a0944
```
http://coliru.stacked-crooked.com/a/2b6397e083f157fb
https://ideone.com/nIrF22

**Comment by schoolo** [comment]
@Bu11etmagnet  that makes sense, thanks for pointing that out, I learned something new.  There is potentially a different problem here then, and that is that when I use a `NiceMock`, I can not use `Mock::VerifyAndClear`. `NiceMock` subclasses my mock object and in its constructor up-casts `this` and registers the up-casted pointer.  The up-casted pointer doesn't have to be the same as the address of the `NiceMock` object, by virtue of the defined behavior of `ImplicitCast_()`.  Therefore, the interface `Mock::VerifyAndClear(<address_of_mock_object>)` doesn't seem to be well-defined for constructs like `NiceMock` and maybe `StrictMock`.  Unless I write a wrapper to up-cast my `NiceMock` whenever I want to use this interface.

Note that I believe I found the reason for the cast resulting in a different pointer value.  While my mock object has no virtual functions( therefore not requiring a pointer to the vtable ), `NiceMock` has at least a virtual destructor, and therefore has a pointer to the vtable as its first data member.  When I upcast its address to a mock_object pointer, I get the memory address `&nice_mock + sizeof(ptr_to_vtable)`.

Maybe it is worth adding to the documentation that using `NiceMock` only works on mock objects with virtual functions, or requiring a virtual destructor in any mock object that uses NiceMock.  Or, since I see that the Cheat Sheet does say that virtual destructors are required when mocking a ["normal class"](https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md#mocking-a-normal-class), maybe it is worth putting a reminder in [this section](https://github.com/google/googletest/blob/master/googlemock/docs/CookBook.md#mocking-nonvirtual-methods) , since in my opinion it gives the impression that a totally non-virtual mock is OK.
